### PR TITLE
FAR-231: Create K6 for loading testing Verdaccio

### PR DESCRIPTION
**Pull request by seanmadi**
Load tests created as K6 scripts. 

Run commands from `verdaccio/load-tests`.

Numbers can be adjusted within the scripts. But there is an upper limit for the number of users, as Verdaccio implements rate limiting at 5000 requests every 2 minutes. Based on my math of how long the requests take, you can go up to about *611* users (at the time of my testing, for me, full requests took an avg of 7.33s, so `5000 / 60 * 7.33 = 611.66 users`).

When Dexter was running a script that publishes `marking-review-engine` every 30 seconds, alongside the Load Test script in this PR, he was able to replicate the issue we were concerned about within about 9 minutes.

https://servicetitan.atlassian.net/browse/FAR-1231

----

### Setup

Install K6

```bash
brew install k6
```

#### Load Test

Tests an average load over a medium period of time.

```bash
k6 run --env VERDACCIO_AUTH_TOKEN={token} scripts/load-test.js
```

#### Stress Test

Tests a high load over a medium period of time.

```bash
k6 run --env VERDACCIO_AUTH_TOKEN={token} scripts/stress-test.js
```

#### Spike Test

Tests a very high load over a very short period of time.

```bash
k6 run --env VERDACCIO_AUTH_TOKEN={token} scripts/spike-test.js
```

#### Soak Test

Tests an average load over a long period of time.

```bash
k6 run --env VERDACCIO_AUTH_TOKEN={token} scripts/soak-test.js
```

### CODE REVIEW

**Review by dextersealy — COMMENTED** [review]
Do we need `verdaccio/load-tests/Dockerfile`?

File: verdaccio/load-tests/scripts/soak-test.js [hunk 1]
```diff
@@ -0,0 +1,23 @@
+/**
+ * Spike test with the purpose of testing a sudden spike in traffic that quickly dies off.
```
Comment (dextersealy) [inline verdaccio/load-tests/scripts/soak-test.js:2]: Incorrect description

**Comment by seanmadi** [comment]
> Do we need `verdaccio/load-tests/Dockerfile`?

I don't think so; removed. I am hoping I can use the official k6 image in the github action.

File: verdaccio/load-tests/scripts/load-test.js [hunk 2]
```diff
@@ -0,0 +1,26 @@
+/**
+ * Load test with the purpose of to determine how Verdaccio responses under an average load.
+ *
+ * Ramps up the number of users over 1 minute, then keeps the load stable for 20 mins,
+ * and then ramps down over 1 min.
```
Comment (dextersealy) [inline verdaccio/load-tests/scripts/load-test.js:5]: Comment doesn't match actual numbers. Probably better to drop the comment altogether.

File: verdaccio/load-tests/scripts/soak-test.js [hunk 3]
```diff
@@ -0,0 +1,23 @@
+/**
+ * Soak test with the purpose of testing an average load over a long period of time.
+ */
+
+import { sleep } from 'k6';
+import { requestTarball } from './common.js';
+
+const RAMP_DURATION = '5m';
+const STABLE_DURATION = '8h';
+const TARGET = 200;
```
Comment (dextersealy) [inline verdaccio/load-tests/scripts/soak-test.js:10]: In the load test this was called `USERS_COUNT` which I find more informative. Can we use that terminology in all the tests?

File: verdaccio/load-tests/scripts/stress-test.js [hunk 4]
```diff
@@ -0,0 +1,34 @@
+/**
+ * Stress test with the purpose of seeing how Verdaccio handles an increased load.
+ *
+ * Ramps up the number of requests over 1 min, keeps stable for 5 mins,
+ * then repeats 2 times before ramping down.
+ *
+ * Numbers can be increased until the server fails to test the limit.
```
Comment (dextersealy) [inline verdaccio/load-tests/scripts/stress-test.js:7]: Comment doesn't match actual numbers.

**Review by dextersealy — APPROVED** [review]

**Review by kanoshin — APPROVED** [review]

**Pull request merged**